test/html.test.js

```diff
@@ -89,6 +89,117 @@ describe('html', function() {
       html.patch(elem, jml);
       assert.equal(elem.outerHTML, '<div id="a1"></div>');
     });
+
+    it('should update an element given JsonML with raw (Markup) instance', function() {
+      const rawStr = new html.raw('foo');
+      assert.equal(rawStr.value, 'foo');
+      assert.equal(rawStr.toString(), 'foo');
+      const elem = document.createElement('div');
+      const jml = ['', rawStr];
+      html.patch(elem, jml);
+      assert.equal(elem.outerHTML, '<div>foo</div>');
+    });
+
+    describe('element attributes', function() {
+      it('should be able to update element attributes', function() {
+        const elem = document.createElement('input');
+        const attrs = {
+          id: 'my-el-1',
+          name: 'my-best-el',
+          className: 'foo bar baz',
+          disabled: true,
+        };
+        const jml = ['', attrs];
+        html.patch(elem, jml);
+        assert.equal(elem.outerHTML, `<input id="${attrs.id}" name="${attrs.name}" class="${attrs.className}" disabled="">`);
+      });
+
+      it('should be able to update element styles through attributes', function() {
+        const elem = document.createElement('input');
+        // NOTE: JsonML style attribute must provide the literal inline CSS string value.
+        // An object of DOM style properties (e.g. marginTop) will not be applied as expected.
+        const attrs = {
+          style: 'color: blue; margin-top: 25px;',
+        };
+        const jml = ['', attrs];
+        html.patch(elem, jml);
+        assert.equal(elem.outerHTML, `<input style="${attrs.style}">`);
+      });
+
+      it('should be able to update set event handlers through attributes', function(done) {
+        const elem = document.createElement('input');
+        // NOTE: event names must be lowercase. camelCase will currently fail
+        // click to complete the test otherwise the test will fail due to timeout
+        html.patch(elem, ['', { onclick: () => done() }]);
+        elem.dispatchEvent(new document.defaultView.Event('click'));
+      });
+    });
+
+    describe('tables', function() {
+      it('should attempt to append a TD cell to by creating a tbody', function() {
+        const elem = document.createElement('table');
+        const jml = ['',
+          ['td', 'rogue table cell'],
+        ];
+        html.patch(elem, jml);
+        assert.equal(elem.outerHTML, '<table><tbody><td>rogue table cell</td></tbody></table>');
+      });
```

Comment:
The code that supports this is a bit odd. It's buggy as it stands. The `tbody` should also include a row. However, it would be simpler to just disallow (throw an error) appending an orphaned table cell to a table without any idea of where it ought to go.

Still, I've added tests based on the current behavior to help illustrate what the library is currently doing, if nothing else. I'm not sure what to do about it other than immediately file one or more issues.

Reply:
Filed https://github.com/CondeNast/jsonml.js/issues/27 to handle these